src/simple_graph.py

```diff
@@ -0,0 +1,200 @@
+        else:
+            self._nodes.update({n1: [(n2, weight)]})
+        self.add_node(n2)
+
+    def del_node(self, n):
+        """Delete node from list and remove any reference to it in other nodes."""
+        if self.has_node(n):
+            del self._nodes[n]
+            for node in self._nodes:
+                for edge in self._nodes[node]:
+                    if edge[0] == n:
+                        self._nodes[node].remove(edge)
+        else:
+            raise(KeyError)
+
+    def del_edge(self, n1, n2):
+        """Remove edge from list."""
+        for edge in self._nodes[n1]:
+            if n2 in edge:
+                self._nodes[n1].remove(edge)
+                break
+        else:
+            raise (ValueError)
+
+    def has_node(self, n):
+        """.Check if node is in graph."""
+        return n in self._nodes
+
+    def neighbors(self, n):
+        """Return list of neighbors nodes to node n."""
+        return [edge[0] for edge in self._nodes[n]]
+
+    def adjacent(self, n1, n2):
+        """Return True if n1 and n2 are adjacent to each other."""
+        if not self.has_node(n1) or not self.has_node(n2):
+            raise(KeyError)
+        for edge in self._nodes[n1]:
+            if n2 in edge[0]:
+                return True
+        return False
+
+    def depth_first_traversal(self, start):
+        """Launch a dfs search, exploring all nodes."""
+        s = Stack()
+        self._visited = []
+        self._explore(start, s)
+        return self._visited
+
+    def _explore(self, node, stack):
+        """Explore a given node, updated visited and stack and calls itself with a new unvisited node."""
+        # import pdb; pdb.set_trace()
+        self._visited.append(node)
+        if self.neighbors(node):
+            stack.push(node)
+            for node in self.neighbors(node):
+                if node not in self._visited:
+                    self._explore(node, stack)
+        return stack
+
+    def breadth_first_traversal(self, start):
+        """Launch a dfs search, exploring all nodes."""
+        q = Queue()
+        self._visited = []
+        self._explore_bfs(start, q)
+        return self._visited
+
+    def _explore_bfs(self, node, queue):
+        if node not in self._visited:
+            self._visited.append(node)
+        if self.neighbors(node):
+            for neighbor in self.neighbors(node):
+                if neighbor not in self._visited:
+                    self._visited.append(neighbor)
+                    queue.enqueue(neighbor)
+            while len(queue):
+                self._explore_bfs(queue.dequeue(), queue)
+        return queue
+
+    def dijkstras(self, start, end):
+        """Implementation of dykstras shortest path algorithm."""
+        if start not in self.nodes():
+            raise TypeError('the start node is not in the graph')
+        if end not in self.nodes():
+            raise TypeError('the end node is not in the graph')
```

Comment:
Good idea thinking about these cases.  I believe a ValueError would be more correct here, assuming that start and end are nodes, but not in the graph.